core/src/kura.rs

```diff
@@ -220,36 +193,28 @@ impl Kura {
     }
 
     #[allow(clippy::expect_used, clippy::cognitive_complexity, clippy::panic)]
-    fn kura_recieve_blocks_thread(
-        weak_kura_reference: &Weak<Kura>,
+    fn kura_recieve_blocks_loop(
+        kura: &Kura,
         mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
     ) {
+        let mut is_closed = false;
         loop {
-            // Here we essentially check if the thread should keep running.
-            // If all other references to `Kura` have been dropped, then
-            // the thread associated with that `Kura` instance should exit.
-            let strong_kura_reference = if let Some(kura_arc) = weak_kura_reference.upgrade() {
-                kura_arc
-            } else {
-                info!("Kura block thread is shutting down");
-                return;
-            };
+            let mut block_reciever_guard = kura
+                .block_reciever
+                .lock()
+                .expect("able to lock kura recieve block mutex");
```

Comment:
The `block_reciever` lock was acquired every loop iteration as a consequence of the `weak_kura_referenc`. Since you now have an owned reference now you can pull this statement out of the loop.

Reply:
It's a bit pedantic of a suggestion but you seem to care about being optimal.

Reply:
Good point, i overlooked that the thread is the only user of receiver so we can freely take the lock.